### PR TITLE
support indexes on vertices

### PR DESCRIPTION
**Pull request by mourednik**
* Allows creation of indexes on any class
* Implements StepStrategy vendor optimization to use indexes in traversal
* Currently the StepStrategy will only find and use indexes that are on class V.

### CODE REVIEW

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/traversal/step/sideEffect/OrientGraphStep.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                    .iterator();
+        }
+    }
+
+    //TODO: indexed edges
+    private Iterator<? extends Edge> edges() {
+        final OrientGraph graph = (OrientGraph) this.getTraversal().getGraph().get();
+        return this.iteratorList(graph.edges());
+    }
+
```
Comment (mourednik) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/traversal/step/sideEffect/OrientGraphStep.java]: The following methods are basically copied from `TInkerGraphStep.java`

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/traversal/strategy/optimization/OrientGraphStepStrategy.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package org.apache.tinkerpop.gremlin.orientdb.traversal.strategy.optimization;
```
Comment (mourednik) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/traversal/strategy/optimization/OrientGraphStepStrategy.java]: This is all copied from `TInkerGraphStepStrategy.java`

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java [hunk 3]
```diff
@@ -52,11 +62,10 @@ public static OrientGraph open(final Configuration configuration) {
     public OrientGraph(Configuration config) {
         this.url = config.getString(CONFIG_URL, "memory:test-" + Math.random());
         this.database = getDatabase(url,
-            config.getString(CONFIG_USER, "admin"),
-            config.getString(CONFIG_PASS, "admin"),
-            config.getBoolean(CONFIG_CREATE, true),
-            config.getBoolean(CONFIG_OPEN, true));
-        makeActive();
+                config.getString(CONFIG_USER, "admin"),
```
Comment (mpollmeier) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: 8 spaces for indent?

File: driver/src/test/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraphIndexTests.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        config.setProperty("type", "UNIQUE");
+        config.setProperty("keytype", OType.STRING);
+        graph.createIndex(keyName, Vertex.class, config);
+
+        Assert.assertEquals(graph.getIndexedKeys(Vertex.class), new HashSet<String>(Collections.singletonList(keyName)));
+
+        graph.addVertex(T.label, className, keyName, value);
+
+        // This test doesn't check that the following traversal hit the index.
+        // Only verified by println debugging in OrientGraphStep.
```
Comment (mourednik) [inline driver/src/test/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraphIndexTests.java]: I suppose that there should be a test for the traversals use of indexing in a `OrientGraphStepTest` suite. Haven't figured out a way to do that yet. It seems that TinkerGraph doesn't test it either.

File: driver/src/test/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraphStepStrategyTest.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package org.apache.tinkerpop.gremlin.orientdb;
```
Comment (mourednik) [inline driver/src/test/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraphStepStrategyTest.java]: This is all copied from `TInkerGraphStepStrategyTest.java`

**Comment by mpollmeier** [comment]
As discussed, LGTM

File: driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java (hunk 3 above)
Comment (mourednik) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: It's a continuation of a previous line. Makes sense here. Automatically formatted

Comment (mourednik) [inline driver/src/main/java/org/apache/tinkerpop/gremlin/orientdb/OrientGraph.java]: I'll change the continuation indent to 4 spaces

**Pull request merged**